File: composeApp/src/androidMain/kotlin/com/ivangarzab/kluvs/ui/books/BookCoverPlaceholder.kt
package com.ivangarzab.kluvs.ui.books

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.width
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.tooling.preview.PreviewLightDark
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.ivangarzab.kluvs.theme.KluvsTheme
import com.ivangarzab.kluvs.theme.lightBar
import com.ivangarzab.kluvs.theme.warmDarkBar
import kotlin.math.sqrt

/**
 * The design system's "no cover available" fallback: a tessellating hexagon
 * hive grid on the [warmDarkBar]/[lightBar] surface. See design-system/docs/book-cover.md.
 *
 * Pointy-top hexagon tiling where [hexWidth] is the vertex-to-vertex column spacing
 * (matches the web token's 28px tile width at book-cover--md scale).
 */
@Composable
fun BookCoverPlaceholder(
    modifier: Modifier = Modifier,
    hexWidth: Dp = 14.dp
) {
    val strokeColor = MaterialTheme.colorScheme.outline
    val backgroundColor = if (isSystemInDarkTheme()) warmDarkBar else lightBar

    Canvas(
        modifier = modifier
            .fillMaxSize()
            .background(backgroundColor)
    ) {
        val hexWidthPx = hexWidth.toPx()
        val radius = hexWidthPx / sqrt(3f)
        val hexHeight = radius * 2f
        val rowSpacing = hexHeight * 0.75f

        val path = Path()
        var row = 0
        var y = -hexHeight
        while (y < size.height + hexHeight) {
            val xOffset = if (row % 2 == 1) hexWidthPx / 2f else 0f
            var x = -hexWidthPx + xOffset
            while (x < size.width + hexWidthPx) {
                addHexagon(path, Offset(x, y), radius)
                x += hexWidthPx
            }
            y += rowSpacing
            row++
        }

        drawPath(path, color = strokeColor, style = Stroke(width = 1.5.dp.toPx()))
    }
}

/** Appends a pointy-top hexagon outline centered at [center] with circumradius [radius]. */
private fun addHexagon(path: Path, center: Offset, radius: Float) {
    for (i in 0..5) {
        val angleDeg = -90.0 + 60.0 * i
        val angleRad = Math.toRadians(angleDeg)
        val point = Offset(
            x = center.x + (radius * kotlin.math.cos(angleRad)).toFloat(),
            y = center.y + (radius * kotlin.math.sin(angleRad)).toFloat()
        )
        if (i == 0) path.moveTo(point.x, point.y) else path.lineTo(point.x, point.y)
    }
    path.close()
}

@PreviewLightDark
@Composable
fun Preview_BookCoverPlaceholder() = KluvsTheme {
    Box(modifier = Modifier.background(MaterialTheme.colorScheme.background)) {
        BookCoverPlaceholder(
            modifier = Modifier
                .width(120.dp)
                .aspectRatio(2f / 3f)
        )
    }
}

File: iosApp/iosApp/Views/Books/BookCoverPlaceholder.swift
//
//  BookCoverPlaceholder.swift
//  iosApp
//
import SwiftUI

/// The design system's "no cover available" fallback: a tessellating hexagon
/// hive grid on the warmDarkBar/lightBar surface. See design-system/docs/book-cover.md.
///
/// Pointy-top hexagon tiling where `hexWidth` is the vertex-to-vertex column spacing
/// (matches the web token's 28px tile width at book-cover--md scale).
struct BookCoverPlaceholder: View {
    var hexWidth: CGFloat = 14

    @Environment(\.colorScheme) private var colorScheme

    private var backgroundColor: Color { colorScheme == .dark ? .warmDarkBar : .lightBar }
    private var strokeColor: Color { colorScheme == .dark ? .warmDarkCard2 : .lightDivider }

    var body: some View {
        Canvas { context, size in
            let radius = hexWidth / sqrt(3)
            let hexHeight = radius * 2
            let rowSpacing = hexHeight * 0.75

            var path = Path()
            var row = 0
            var y = -hexHeight
            while y < size.height + hexHeight {
                let xOffset: CGFloat = row % 2 == 1 ? hexWidth / 2 : 0
                var x = -hexWidth + xOffset
                while x < size.width + hexWidth {
                    addHexagon(to: &path, center: CGPoint(x: x, y: y), radius: radius)
                    x += hexWidth
                }
                y += rowSpacing
                row += 1
            }

            context.stroke(path, with: .color(strokeColor), lineWidth: 1.5)
        }
        .background(backgroundColor)
    }

    /// Appends a pointy-top hexagon outline centered at `center` with circumradius `radius`.
    private func addHexagon(to path: inout Path, center: CGPoint, radius: CGFloat) {
        for i in 0...5 {
            let angleDeg = -90.0 + 60.0 * Double(i)
            let angleRad = angleDeg * .pi / 180.0
            let point = CGPoint(
                x: center.x + radius * cos(angleRad),
                y: center.y + radius * sin(angleRad)
            )
            if i == 0 {
                path.move(to: point)
            } else {
                path.addLine(to: point)
            }
        }
        path.closeSubpath()
    }
}

#Preview {
    BookCoverPlaceholder()
        .frame(width: 120, height: 180)
}
